src/utils/hotcode-agent/src/main/java/io/bellsoft/hotcode/agent/HotCodeAgent.java

```
@@ -0,0 +1,96 @@
+                configuration.samplingInterval, configuration.profilingDuration));
+        if (configuration.profilingPeriod.isZero()) {
+            service.schedule(
+                    task,
+                    configuration.profilingDelay.toMillis(),
+                    TimeUnit.MILLISECONDS);
+        } else {
+            service.scheduleAtFixedRate(
+                    task,
+                    configuration.profilingDelay.toMillis(),
+                    configuration.profilingPeriod.toMillis(),
+                    TimeUnit.MILLISECONDS);
+        }
+    }
+
+    private static void parseArgs(String argumentString, Properties properties) {
+        if (argumentString != null) {
+            var arguments = argumentString.split(",");
+            for (var argument : arguments) {
+                int idx = argument.indexOf('=');
+                if (idx >= 0) {
+                    var key = argument.substring(0, idx);
+                    var value = argument.substring(idx + 1);
+                    properties.put(key, value);
+                } else {
+                    properties.put(argument, "");
+                }
+            }
+        }
+    }
+
+    private class ProfileAndOptimize implements Runnable {
+        private ProfilingTask profilingTask;
+
+        public ProfileAndOptimize(ProfilingTask profilingTask) {
+            this.profilingTask = profilingTask;
+        }
+
+        @Override
+        public void run() {
+            try {
+                var hotMethods = profilingTask.call().getTop();
+                var directives = CompilerDirectives.build(hotMethods);
+                directivesControl.replace(directives);
+            } catch (Throwable e) {
+                LOGGER.log(Level.SEVERE,
+                        "an error occurred during the profiling task execution, further profiling is cancelled", e);
+                throw new RuntimeException(e);
```

It makes sense to log an exception here, or it will be silently swallowed.

---

Do you mean some extra logging? (`e` is passed to `log` here)

---

Oops, sorry, I didn't notice `e` is already logged.